### ISSUE TITLE
How do I upload files with Octokit to Github signed with a verified label?

### ISSUE THREAD

**Issue by LukeXF** [octokit/rest.js#1666]
Using [Octokit REST API](https://github.com/octokit/rest.js), I have authentication working and I'm trying to upload using [createOrUpdateFile](https://octokit.github.io/rest.js/v17#repos-create-or-update-file), which is also working. I have my commits been made by the GitHub App's app currently. 

But when viewing all commits by that app, it just displays this message below:

[![enter image description here][1]][1]

I want to have my commits to run through a user that I've made `MerlinPanelBot`. So I can view all commits made by that user, and have more detailed analytics on the user's profile (MerlinPanelBot) and for insights in each project the code commits.

###Is there an easy way to use my new user `MerlinPanelBot` in my code AND have verified commits?

Perhaps I need to look at using oAuth and/or getting a Personal Access Token to run directly as the user? Hoping it's a simple fix.

---
# Progress so far:

1. When not providing committer, it will commit a verified GPG Key with the Github App's name `merlin-panel-deploy-bot`. This is good, but I want it to run through a new user I've created called `MerlinPanelBot`.

2. I set the committer to my own account `LukeXF` to test that it works. It did but there is no verified GPG label on my commit.
```
committer: {
	name: "LukeXF",
	email: '<my email (redacted)>'
},
```
3. I tried changing the email address in committer to the MerlinPanelBot (user), which when I ran the code threw an unauthorized error.

4. Adding a PEM file for MerlinPanelBot on the GitHub App's page for the user MerlinPanelBot (as shown below) allows my code to run, but still no verified GPG label.
[![enter image description here][2]][2]

Bullet points visualised:
[![enter image description here][3]][3]

---

# Code

Even though it took a long time to get authentication, reading, buffering and sending files with Github/Octokit working, my code is actually quite simple, as follows:

Route File:
```js
const installationAccessToken = await GithubClass.getInstallationAccessToken('repoExample');
const octokit = new Octokit({
	auth: `token ${installationAccessToken}`
});
const fromGithub = await GithubClass.sendFile(octokit, 'repoExample', 'MERLIN.md', req.body.lineToAdd, req.userId);
return res.send(fromGithub);
```

Built octokit auth:
```js
// as explained above I have two PEM files for my two account used in testing.
const pemFile = path.resolve(__dirname, '../app.2020-02-02.private-key.pem');

async getInstallationAccessToken(repo) {
	const privateKey = fs.readFileSync(pemFile);
	const app = new App({id: '52777', privateKey});
	const jwt = app.getSignedJsonWebToken();

	// Example of using authenticated app to GET an individual installation
	// https://developer.github.com/v3/apps/#find-repository-installation
	const {data} = await request("GET /repos/:owner/:repo/installation", {
		owner: "element-softworks",
		repo,
		headers: {
			authorization: `Bearer ${jwt}`,
			accept: "application/vnd.github.machine-man-preview+json"
		}
	});
	// contains the installation id necessary to authenticate as an installation
	const installationId = data.id;

	// return installationId;
	return await app.getInstallationAccessToken({
		installationId,
	});
},
```
Function to update file:
```js
async sendFile(octokit, repo, path, lineToAdd, userId) {
	const repos = await octokit.repos.getContents({
		owner: 'element-softworks',
		repo,
		path
	});

	const {sha} = repos.data;
	let file = Buffer.from(repos.data.content, 'base64').toString();

	file += `
${lineToAdd} by ${userId}`;

	return await octokit.repos.createOrUpdateFile({
		owner: 'element-softworks',
		repo,
		path,
		message: `Updating product ${lineToAdd} by ${userId}`,
		content: Buffer.from(file).toString('base64'),
		committer: {
			name: "Luke",
			email: '<my email (redacted)>'
		},
		sha
	})
}
```

  [1]: https://i.stack.imgur.com/x2v5A.png
  [2]: https://i.stack.imgur.com/GSFgY.png
  [3]: https://i.stack.imgur.com/ovGec.png

**Comment by gr2m** [comment]
I don't think that it's possible to sign commits using the [`octokit.repos.createOrUpdateFile()`](https://octokit.github.io/rest.js/v17/#repos-create-or-update-file) method.

I'll look into it, there might be an alternative way

**Comment by gr2m** [comment]
You can use the https://developer.github.com/v3/git/commits/#create-a-commit API in order to create verified commits. The process is more involved though, because you need to create a git tree first: https://octokit.github.io/rest.js/v17#git-create-tree

Could you give that a try? Once you figure out the pieces, this would make a great Octokit plugin for others to use 👍

**Comment by LukeXF** [comment]
> I don't think that it's possible to sign commits using the [`octokit.repos.createOrUpdateFile()`](https://octokit.github.io/rest.js/v17/#repos-create-or-update-file) method.
> 
> I'll look into it, there might be an alternative way

I have commits signed using the bot which is great, so I know it is possible. I think it's something to do with the authentication on why it's not working when using the user account.

Thank you for your help so far buddy @gr2m

**Comment by gr2m** [comment]
The alternative way is to use the git APIs as I described in [my comment above](https://github.com/octokit/rest.js/issues/1666#issuecomment-611715098)

**Comment by gr2m** [comment]
Sorry this should not have been flagged as stale

**Comment by LukeXF** [comment]
Yeah, I've still not got this working, right now it's using the bot rather than the Verified user (just to keep you updated).
